src/hooks/Circles/utils/optimisticallyUpdatePosts.ts

```diff
@@ -4,65 +4,62 @@ import type {
   PostDetailsPostQueryResponse,
 } from '../useInfinitePosts';
 import { Post } from '../types';
+import forEach from 'lodash/forEach';
+import { cloneDeep } from 'lodash';
 
 type UpdatePostsConfig = {
   queryClient: QueryClient;
-  postId: string;
-  updatePost: (post: Post) => Post;
+  id: string;
+  transformFn: (post: Post) => Post;
 };
 
 export const optimisticallyUpdatePosts = ({
   queryClient,
-  postId,
-  updatePost,
+  id,
+  transformFn,
 }: UpdatePostsConfig) => {
   queryClient.setQueryData<InfinitePostsData>(['posts'], (currentData) => {
-    const edge = findEdgeById(currentData!, postId);
-
-    if (!edge.node) {
-      return currentData!;
-    }
-
-    edge.node = updatePost(edge.node);
-
-    return currentData!;
+    const newData = cloneDeep(currentData);
+    forEach(newData?.pages, (page) => {
+      transformEdgesById(page.postsV2?.edges, id, transformFn);
+    });
+    return newData!;
   });
 
-  const visited: Post[] = [];
   queryClient
     .getQueryCache()
     .findAll({ queryKey: ['postDetails'] })
     .map((query) => {
-      query?.setData((currentData: PostDetailsPostQueryResponse) => {
-        if (!currentData) {
-          return currentData!;
-        }
-
-        const posts = extractPosts({ node: currentData.post });
-
-        const node = posts.find((post) => post.node?.id === postId);
-
-        if (node?.node && visited.indexOf(node.node) === -1) {
-          visited.push(node.node);
-          node.node = updatePost(node.node);
-        }
-
-        return currentData;
-      });
+      query?.setData(
+        (currentData: PostDetailsPostQueryResponse | undefined) => {
+          const newData = cloneDeep(currentData);
+          if (newData?.post.id === id) {
+            newData.post = transformFn(newData.post);
+            return newData;
+          }
+          if (newData?.post.replies?.edges) {
+            transformEdgesById(newData.post.replies?.edges, id, transformFn);
+          }
+          return newData;
+        },
+      );
     });
 };
 
-const extractPosts = (post: { node: Post }): { node: Post }[] => [
-  post,
-  ...((post.node?.replies?.edges.map(extractPosts).flat() ?? []) as {
-    node: Post;
-  }[]),
-];
-
-const findEdgeById = (currentData: InfinitePostsData, id: string) =>
-  currentData?.pages
-    .map((page) => page.postsV2.edges)
-    .flat()
-    .find(({ node }) => node.id === id) ?? {
-    node: undefined,
-  };
+const transformEdgesById = (
+  edges: { node: Post }[],
+  id: string,
+  transformFn: (post: Post) => Post,
+) => {
+  if (!edges) {
+    return;
+  }
+  forEach(edges, (edge) => {
+    if (edge.node.id === id) {
+      edge.node = transformFn(edge.node);
+    }
+    if (edge.node.replies?.edges) {
+      transformEdgesById(edge.node.replies?.edges, id, transformFn);
```

Comment:
Infinite recursion is usually dangerous but the dataset here should never contain a circular reference and be fairly thin.

Reply:
Good call out - can we add some early exit logic just in case?  e.g. Is there a number that'd be obviously too large?

Reply:
Yeah, we could add some bail out logic just in case. I'll update it.

Reply:
Done